### DISCUSSION TITLE
How to use Forward function to calculate a zero order sweep.

### DISCUSSION THREAD

**Discussion by ghost** [coin-or/CppAD#128] [Q&A] [answered]
Hi, I want to use Forward function to calculate zero order sweep. 
```
CppAD::ADFun<double> getStateDerivative;
CPPAD_TESTVECTOR(double) x(31);
return getStateDerivative.Forward(size_t(0), x);
```
But the compiler reports
```
error: no matching function for call to ‘CppAD::ADFun<double, double>::Forward(size_t, CppAD::vector<double>&)
```
I have tried different ADFun with different Base type and different `xq` of different BaseVector type but it still does not work.

**Comment by bradbell** [comment]
I created the program
```
int main(void)
{
    CppAD::ADFun<double> getStateDerivative;
    CPPAD_TESTVECTOR(double) x(31), y(31);
    y = getStateDerivative.Forward(size_t(0), x);
    return 0;
}
```
It compiled fine with the following command
```
g++ -g -I $HOME/repo/cppad.git/include temp.cpp -o temp
```
The message I got when I tried to run the program was
```
cppad-20220201 error from a known source:
Forward(q, xq): xq.size() is not equal n or n*(q+1)
Error detected by false result for
    size_t(xq.size()) == n || size_t(xq.size()) == n*(q+1)
at line 119 in the file 
    /home/bradbell/repo/cppad.git/include/cppad/core/forward/forward.hpp
temp: /home/bradbell/repo/cppad.git/include/cppad/utility/error_handler.hpp:206: static void CppAD::ErrorHandler::Default(bool, int, const char*, const char*, const char*): Assertion `false' failed.
environment: line 13: 1211220 Aborted                 (core dumped) ./temp
```
This is because as defined the function getStateDerivative is empty.

**Comment by bradbell** [comment]
@xeonz1 Did the message above answer your quesiton ? If so, would you please close this issue.

**Comment by ghost** [comment, marked as answer]
> @xeonz1 Did the message above answer your quesiton ? If so, would you please close this issue.

@bradbell Thanks for the reply. I have solved the problem above. However, I am running into another problem about codegen. In the generateSource part of zero order forward mode, It reports
> GreaterThanZero cannot be called for non-parameters.
I debugged for a while and found this exception happens when GreaterThanZero is called in 'forward_levv_op_0'. Can you give me some instruction?

**Reply by bradbell** [reply]
> 
@xeonz1 
Would you please ask about this a separate question ```GreaterThanZero cannot be called for non-parameters ?``` on
https://github.com/coin-or/CppAD/discussions/categories/q-a

**Reply by ghost** [reply]
I found the solution. I used eigen inverse for a 6*6 square matrix, which involves ifelse branches(or something similar).  Therefore I used another method to calculate the inverse and everything works.

**Comment by bradbell** [comment]
Converting to a discussion from issue #127
